### DISCUSSION TITLE
Using openfast with MLIfe for fatigue and ccyles to failure

### DISCUSSION THREAD

**Discussion by lt72884** [OpenFAST/openfast#1691] [Q&A]
I have now completed the tasks of creating a turbsim.bts file that works. I have tested the .bts file in my 5MW_LAND_DLL_WTURB.fst and it works great. Thank you so much for all the help. I am very grateful.

Now that i have the one outb file and can view it in matlab and produce the plots. I would like to use MLife next for fatigue

I have installed Mlife and i have the documentation. What do i need to do specifically in Mlife to have it read the .outb file and calculate the fatigue and cycles to failure. 

1. Do i need to have any of the mlife files in the 5MW_LAND_DLL_WTRUB folder?
2.Do i need to edit any files and if so, what are the file extension endings fo those files?
3. What code command do i need to run in matlab to use mlife and calculate values?
4. Once values are calculated, how do i view them?
5. Is there by chance a video or teams meeting or demonstration on Mlife that could help me?

thanks for all the help thus far. It is really appreciated.

**Comment by jjonkman** [comment]
Dear @lt72884,

I'm not aware of a video or workshop tutorial stepping through the details of setting up and executing MLife post-processing.  The MLife user and theory documentation and sample models from the MLife archive (https://www.nrel.gov/wind/nwtc/mlife.html) are you best source of information.

Best regards,

**Reply by lt72884** [reply]
ok, some progress has been made.

I created the following:
C:/fatigue/data
C:/fatigue/outputs

I opened the mlife.m file located in the MLife/Source directory. I copied the test01.mlif file from the MLife cert test folder and placed it in the MLife source directory. I changed its name to settingsFile.mlif. I then copied the DLC2.0_1.out to the C:/fatgue/data directory

i then ran the following lines of code

settingsFile = 'settingsFile.mlif'
dataDirectory = 'C:/fatigue/data,
outputsDirectory = 'C:?fatigue/outputs'

[Fatigue, Statistics]= mlife( settingsFile, dataDirectory, outputDirectory );

it then ran and created some fles that i dont understand the output. Matlab also threw the following errors:
'wblcdf' requires Statistics and Machine Learning Toolbox.

Error in compute_windspeed_bins/compute_bins (line 74)
            bins(iBin) = wblcdf(WShi, c, k) - wblcdf(WSlo, c, k) ;

Error in compute_windspeed_bins (line 32)
   [nBins1, binWidth1, binProbabilites1] = compute_bins(0, Fatigue.WSin, Fatigue.WSMaxBinSize);

Error in mlife (line 565)
         Fatigue = compute_windspeed_bins(Fatigue, windChannelMeans,FileInfo.DLCs(1).NumFiles, FileInfo.DLCs(2).NumFiles );

so we are one step closer. What is next to fix this?
How do i use my .outb file generated from open fast IN Mlife to calculate fatigue and etc?

thanks

**Reply by jjonkman** [reply]
Dear @lt72884,

According to the MLife User's Guide, you need the Statistics Toolbox of MATLAB to run MLife within MATLAB.  If you don't have that, you'll need to purchase it or replace the functions supplied by that Toolbox with your own.  Alternatively, you can use the provided precompiled version of MLife.

Best regards,

**Comment by lt72884** [comment]
Ok, I'll look into that. Thanks

Does MLife use the .outb file produced by openfast to calculate the stats and fatigue?

**Reply by jjonkman** [reply]
Yes.  MLife can process one or many OpenFAST time-series output files.

**Reply by ebranlard** [reply]
For simple fatigue analysis (equivalent load), you can also adapt the following python script:
https://github.com/OpenFAST/python-toolbox/blob/dev/pyFAST/postpro/examples/Example_EquivalentLoad.py

**Reply by lt72884** [reply]
ok, great. How do i do that? :)

thanks

**Reply by lt72884** [reply]
ok, something i have noticed.the first link you gave me supplied a version of fatigue.py that does not have find range count. as your second link shows. The first link you gave me for "equivalent load" i downloaded all the files and the fatigue.py in that is not the same as the second. can i just replace it?

thanks

**Reply by ebranlard** [reply]
yes you can replace it.

**Reply by mellow753** [reply]
> Yes. MLife can process one or many OpenFAST time-series output files.

Hi, I have a few questions about MLife: 

1. Can MLife process the time history stored into .xlsx file? For example, two columns ([Time] and [Singla]) in excel file.
2. For one time-series, can MLife output its cycle-amplitudes and cycle-means of rainflow counting?
3. The theory manual indicates MLife uses "one-pass" rainflow (i.e. Algorithm II in Downing and Soice 1982). Does the rainflow code of Algorithm II is in Source\Rainflow\*.m?

Thanks

**Reply by jjonkman** [reply]
Dear @Wying301,

Here are my responses:
1. MLife can write Excel-based output files, but is not set up to read Excel-based time-series files.  I would suggest writing your Excel-based time-series into a file format that MLife can read, such as ascii (`FileFormat` = 1) or OpenFAST binary (`FileFormat` = 2) format, or changing the MLife source code to read Excel-based time-series files.
2. While the raw data computed by the rainflow counting algorithm is available within MATLAB, MLife does not have a feature that will write this data to a file.  But I'm sure this is possible with a small source-code change.
3. As mention in the MLife User's Guide, the rainflow cycle counting source code is written in C++ and compiled as a MEX function for calling within MATLAB.

Best regards,

**Comment by lt72884** [comment]
> For simple fatigue analysis (equivalent load), you can also adapt the following python script: https://github.com/OpenFAST/python-toolbox/blob/dev/pyFAST/postpro/examples/Example_EquivalentLoad.py

How do i run this? i do not program at all and have ZERO education in programming. 

i did a little googling and i used powershell to run the script you provided but i get an error about numpy in line 7

 File "C:\fatigue\fatigue.py", line 7, in <module>
    import numpy as np
ModuleNotFoundError: No module named 'numpy'

and this error as well

from pyFAST.input_output import FASTOutputFile
ModuleNotFoundError: No module named 'pyFAST'

for kicks and giggles i ran this from the power shell which completed sucessfully but it still throws the same error

pip install pyFAST

so i have no idea what to do now

**Reply by lt72884** [reply]
after having professor at my univeristy look at the python code for a couple of hours, we found the issue., well about 7 issues. Once we got them fixed i get the following output 

Leq : 1244.1126390380857

from what i understand, this is the DEL or Damage equivalent load, also known as the load equivalent in some texts. 

what are the units of this number and what IS the meaning of this number? i know what it is calculated from, but what is the meaning? is there any way to back out the cycles to failure? 

thanks.

The snippet below is just some comments from the python script for reference
---
Parameters
    ----------
    time : array-like, the time values corresponding to the signal (s)
    signals : array-like, the load signal
    m :    Wohler exponent (default is 3)
    Teq : The equivalent frequency (Default 1Hz)
    nBins : Number of bins in rainflow count histogram
    method: 'rainflow_windap, rainflow_astm, fatpack

    Returns
    -------
    Leq : the equivalent load for given m and Teq

**Reply by jjonkman** [reply]
Dear @lt72884,

The damage-equivalent load is the constant amplitude, constant frequency cyclic load that provides the same fatigue damage as the time-varying load you specify, based on Miner's rule.  The unit of the damage-equivalent load will be the same unit as the time-series load you provide.

For various calculations related to number of cycles to failure and fatigue damage lifetime, see the fatigue calculations presented in the MLife Theory Manual I shared with you previously.  I'm not sure which of these calculations are supported by the Python Toolbox, but Emmanuel can comment on that.

Best regards,

**Reply by ebranlard** [reply]
The function `equivalent_load` from the python toolbox returns a result consistent with equation (30) of the MLife theory manual. 

If you need to perform more advanced calculations of damage, you can get the ranges and cycles from the function `find_range_count` ([here](https://github.com/OpenFAST/python-toolbox/blob/dev/pyFAST/tools/fatigue.py#L99)) and then apply the math given in the MLife theory document. There is currently no wrapper function (equivalent to MLife) to compute the damage from a set of time series, using the Goodman's correction and the Weibul distribution.

**Reply by lt72884** [reply]
will your script provide cycles to failure? also, what units is Leq in?

How do i get the ranges and cycles fron find_range_count? i know i can use print() and see the values of some things.  I know time is 60 seconds and Teq is 60 as well. 

thanks

Ill look at equation 30 for reference.

**Reply by lt72884** [reply]
thank you:) the Leq value i provided above, i do not know wha the units are. the output is just a number so i am not sure what the units are.

I have been reading the theory manual and it has helped alot. I have been doing some calculations by hand as well. My professor showed me the print() function that i can use to see the output of certain things in the python. 

Ill see if i can back out the cycles to failure somehow from the information i have.

**Reply by ebranlard** [reply]
As Jason mention, the equivalent load is returned in the same unit as your input variable. If your input is in [N], your output will be in [N].

**Reply by lt72884** [reply]
ok, thank you. I understand that now. I get what he means by saying "the time series load you provide" 

as for the python question, i will replace the file

Next, in order to run find_range_count (or even the function eq_load_cycles), do i need to change line 10 from the Example_EquivalentLoad.py from 

"from pyFAST.postpro import equivalent_load" to "from pyFAST.postpro import find_range_count" and make sure init.py in pyFAST>postpro has 

from pyFAST.tools.fatigue import find_range_count

listed as well?

If this is not the correct procedure to run find_range_count or eq_load_cycles, would you be willing to share how to run those? thanks

as of right now, the ONLY output im getting is Leq:1244.76543 

when i do get find_range_count to work, and i get the cycles for each bin, is that cycles to failure? or how many cycles it went under? IE 10^6, 10^7, or 10^8? and will it explicitly tell me?

thank you very very much

PS, i may have found an error in your code. The comment at the top of fatigue.py says m is a default value of 3 for the wohler exponent. BUT on line 87 

 DELi = S**m * N / neq
    print(m)
    print(DELi)
    
when i do a print(m) to see the output, it gives me that m=1 and not 3. I also verified it by doing the math by hand and m=1.

[(241.96)(13)]/60 = 52.426 which is the first DELi value listed in the table that is out put to vs code

here is the output of the print(m) and first few rows of print(DELi)

1

52.42614038   0.          96.7867207    0.         120.98340088
   0.          96.7867207    0.          80.65560059   0.
 120.98340088   0.           0.           0.          32.26224023
   0.           0.           0.           0.           0.
  44.36058032   0.          96.7867207    0.           0.
  
  is it possible to change the value of m to 3?

**Reply by lt72884** [reply]
so DEL is NOT the same as Leq then? If i am reading the code correctly (Leq = DELi.sum() ** (1/m)), then Leq = DEL^(1/m) where m = 3 (by default) so in my case my leq is 1244 newtons so if i do the math then DEL = (1244)^3 which is a large number.  This DEL value of (1244)^3 is the load produced under constant amplitude cycles that has the same degree of damage as real irregular loads?

suppose my Neq is 60 meaning "Number of equivalent cycles to failure." So failing at 60 cycles is that failing under the DEL value or the Leq value?

IE, 60 cycles to fialure at 1244 newtons or 60 cycles to fialure with (1244)^3 newtons?

thank you:)

**Reply by ebranlard** [reply]
The script uses `DELi` to mention the comtribution of cycle "i" to the DEL (or `Leq` both mean the same). 

If you want to look at "damage", you need to apply the theory presented in MLife theory manual, and combine the cycles (`S`) and counts (`N`) yourself as currently we don't have a python-script for that.

**Comment by lt72884** [comment]
I THINK i got it. Here are some screenshots:

https://photos.app.goo.gl/Jz8ATyiSQq5qsXQR9
https://photos.app.goo.gl/UfCC2tUf4uxsB3xt7

i then used equation 26 and 30 to validate the math. However, the m value in the script is NOT defaulted to 3 like the comment says, its defaulted to 1. I used excel to test out and verify that it is indeed m=1 and it is. But i can change that value in excel now and get new numbers. 

Can you confirm that the output i show in the images is indeed cycles and load range according to the script comments?

thanks a million guys

**Reply by ebranlard** [reply]
Yes, "S" and "N" in the code corresponds to the cycle ranges and number of cycles.

**Comment by lt72884** [comment]
now i have a theory question of the script. I noticed it is making 100 bins and some of those bins are empty as shown in the provided screen shots. Why are some of them empty and how is it generating these bins? And lastly, what is a bin?

thanks

**Reply by ebranlard** [reply]
Here's a link that explains data binning:  https://en.wikipedia.org/wiki/Data_binning .
In the script, the amplitude ranges of the signals are split into multiple bins and counted according to these bins.

**Comment by lt72884** [comment]
question. line 26 of the script you made

Leq = equivalent_load(df['Time_[s]'], df['RootMyc1_[kN-m]'], m=m)

The RootMyc1 i know is a parameter that i added to the outlist section of my elastodyn.dat file. However, can i change that RootMyc1 in the script to some other parameter? 

second, since the RootMyc1 is Kn-m that means leq is in Kn-m correct?

thanks

**Reply by ebranlard** [reply]
Yes you can use anyother channel present in the pandas dataframe. You can see them using `print(df.keys())`. 
The units are indeed kN-m for that channel, and therefore the same for the damage equivalent load.

**Reply by lt72884** [reply]
Ok, thank you. I will try and find one that is Newton's only. Thank you again for all the help. I'll try the print function to see what the pandas tell me haha

Does the print function you list go in the fatigue.py script?

**Comment by lt72884** [comment]
ok, i have been working on the script and have some good results. I switched the parmeter from RootMyc1 to RootFxb1 so my DEL is in KN and not a moment. 

Reading the theory manual, i am trying to find Neq cycles until failure. I have neq which is a differnt value, according to the manual neq is equivalent counts for time-series j.  neq =feq*T so 1*60. neq=60.

ok, now i know that Dj^st=neq/Neq (eq 25 in theory manual)
is Dj^st the same as DEL? or is Dj^st different? If it is differnet, does your script calculate Dj^st at all? if so, i would like to know where so i can find Neq which is cycles until failure.

If i knew the fixed mean load, and the ultimate load, i could find Neq as well. Does the script calculate  the fixed mean load and ultimate load?

thanks you very much

**Comment by lt72884** [comment]
oh, i forgot to ask, is there any way to change Teq from the default of 1hrz to something else?

Teq : The equivalent period (Default 1, for 1Hz)

the reason i ask, when i look at the graphical output of RootFxb1 in either pydatview or matlab, in a 15 second span of time, there are only 3 periods. But the python script is set to a default of 1. The manual says "You must enter the equivalent frequency (Teq) of the DEL" im trying to find out what that Teq value is.



thanks
